### DISCUSSION TITLE
Errors creating context beans with a factory bean

### DISCUSSION THREAD

**Discussion by tobiasschaefer** [micronaut-projects/micronaut-core#5041] [Q&A]
We first had a context bean like this:

```java
@Factory
class ProcessEngineFactory {
  @Context
  ProcessEngine processEngine(ProcessEngineConfiguration processEngineConfiguration) {
     // create ProcessEngine based on "processEngineConfiguration"
     // [...]
  }
}
```

and later replaced this with

```java
@Factory
class MyFactory {

  protected final ProcessEngineConfiguration processEngineConfiguration;

  public ProcessEngineFactory(ProcessEngineConfiguration processEngineConfiguration) {
        this.processEngineConfiguration = processEngineConfiguration;
  }

  @Context
  ProcessEngine processEngine() {
     // create ProcessEngine based on "processEngineConfiguration"
     // [...]
  }
}
```

which should be the same, right?

Strangely, the change lead randomly to either build success, very long builds or even OutOfMemory on Github Actions. In the matrix build we only observed errors for JDK 8 (JDK 11 and 15 worked) but this could coincidence. We're using the latest Micronaut 2.3.3.

This is the commit that broke it: https://github.com/NovatecConsulting/micronaut-camunda-bpm/commit/da8be6d59ad0b8264949cceed0eb289b828c270c (there are some unrelated changes in there).

This is the commit that fixed it again: https://github.com/NovatecConsulting/micronaut-camunda-bpm/commit/b9ac02feaabf47fa02305d9034f631ebd746d449

It seems that the instantiation of the beans is non-determinstic and that there is some kind of recursion. Depending on the order it either succeeds, takes long or even fails.

I know there is a simple workaround for us but this cost us several hours and others might run into the same issue. Especially, the sporadic failures make it hard to isolate.

Any thoughts on this?

**Comment by graemerocher** [comment]
Most likely cause is some kind of undetected circular reference causing contention in the synchronised map that holds the beans but without a reproducer hard to say

**Reply by tobiasschaefer** [reply]
@graemerocher : I tried to create a simple project to reproduce the error but failed. However,  I managed to reproduce the error in our open source project. It failed twice in six runs.

Get our project:
`git clone https://github.com/NovatecConsulting/micronaut-camunda-bpm.git`

Checkout the failing commit
`git checkout 02098bd`

Switch to Java 8 Zulu because that is known to fail on GitHub Actions. Other combinations might be affected too, but it seems that other Java Versions or Micronaut Runtimes don't produce the error:
`sdk use java 8.0.282-zulu`

Build:
` ./gradlew clean build -Pmicronaut.runtime=netty`

This will sporadically fail with
```
ProcessEngineFactoryLoggerTest > Camunda version is logged on application start() FAILED
    io.micronaut.context.exceptions.BeanInstantiationException at ProcessEngineFactoryLoggerTest.kt:44
        Caused by: io.micronaut.context.exceptions.BeanInstantiationException at ProcessEngineFactoryLoggerTest.kt:44
            Caused by: java.lang.OutOfMemoryError at XNode.java:51
```

I hope this helps you.
